lincheck/src/main/java/org/jetbrains/kotlinx/lincheck/strategy/ManagedStrategyTransformer.java

```diff
@@ -48,80 +46,659 @@
+        @Override
+        public String map(String name) {
+            // TODO: remove this exception check when exceptions in transformable strategies will be supported
+            boolean isJavaUtilException = name.startsWith("java/util/") && name.endsWith("Exception");
+            boolean isFieldUpdater = name.startsWith("java/util/concurrent/atomic/Atomic") && name.endsWith("FieldUpdater");
+            boolean inFunctionPackage = name.startsWith("java/util/function/");
+            if (name.startsWith("java/util/") && !isFieldUpdater && !inFunctionPackage && !isJavaUtilException) name = TRANSFORMED_PACKAGE + name;
+            return name;
+        }
+    }
+
+    /**
+     * Replaces native methods from java/util with stubs
+     */
+    private static class NativeMethodStubTransformer extends MethodVisitor {
+        GeneratorAdapter mv;
+
+        NativeMethodStubTransformer(GeneratorAdapter mv) {
+            super(ASM_API, null);
+            this.mv = mv;
+        }
+
+        @Override
+        public void visitEnd() {
+            mv.visitCode();
+            mv.push(true);  // suppose that we always have CAS for Long
+            mv.returnValue();
+            mv.visitMaxs(1, 0);
+            mv.visitEnd();
+        }
+    }
+
+    /**
+     * Adds invocations of ManagedStrategy methods before indirect reads and writes of shared variables
+     * via Unsafe or VarHandle
+     */
+    private class IndirectSharedVariableAccessMethodTransformer extends ManagedStrategyMethodVisitor {
+        IndirectSharedVariableAccessMethodTransformer(String methodName, GeneratorAdapter mv) {
+            super(methodName, mv);
+        }
+
+        @Override
+        public void visitMethodInsn(int opcode, String owner, String name, String desc, boolean itf) {
+            if (isUnsafe(owner) || isVarHandle(owner) || isAtomicFieldUpdater(owner)) {
+                String lowerName = name.toLowerCase();;
+                if (lowerName.contains("set") || lowerName.contains("swap") || lowerName.contains("put")) {
+                    invokeBeforeSharedVariableWrite();
+                }
+                if (lowerName.contains("get")) {
+                    invokeBeforeSharedVariableRead();
+                }
+            }
+
+            mv.visitMethodInsn(opcode, owner, name, desc, itf);
+        }
+
+        private boolean isUnsafe(String owner) {
+            return owner.equals("sun/misc/Unsafe");
+        }
+
+        private boolean isVarHandle(String owner) {
+            return owner.equals("java/lang/invoke/VarHandle");
+        }
+
+        private boolean isAtomicFieldUpdater(String owner) {
+            return owner.endsWith("FieldUpdater") && owner.contains("Atomic");
+        }
+    }
+
+    /**
+     * Adds invocations of ManagedStrategy methods before direct reads and writes of shared variables
+     */
+    private class SharedVariableAccessMethodTransformer extends ManagedStrategyMethodVisitor {
```

Comment:
add the logic from `IndirectSharedVariableAccessMethodTransformer` into this method transformer